src/sentry/tasks/unmerge.py

```diff
@@ -0,0 +1,468 @@
+            # ingestion logic (but actually represent a more accurate value.)
+            # See GH-5289 for more details.
+            for value, (times_seen, first_seen, last_seen) in values.items():
+                instance, created = GroupTagValue.objects.get_or_create(
+                    project_id=project.id,
+                    group_id=group_id,
+                    key=key,
+                    value=value,
+                    defaults={
+                        'first_seen': first_seen,
+                        'last_seen': last_seen,
+                        'times_seen': times_seen,
+                    },
+                )
+
+                if not created:
+                    instance.update(
+                        first_seen=first_seen,
+                        times_seen=F('times_seen') + times_seen,
+                    )
+
+
+def collect_release_data(project, events):
+    results = {}
+
+    for event in events:
+        release = event.get_tag('sentry:release')
+
+        # TODO: Double check this!
+        if not release:
+            continue
+
+        # XXX: It's not really clear what the canonical source is for
+        # environment between the tag and the data attribute, but I'm going
+        # with data attribute for now. Right now it seems like they are
+        # intended to both be present and the same value, but I'm not really
+        # sure that has always been the case for existing values.
+        # NOTE: ``GroupRelease.environment`` is not nullable, but an empty
+        # string is OK.
+        environment = event.data.get('environment', '')
+
+        key = (
+            event.group_id,
+            environment,
+            Release.objects.get(  # XXX: This lookup should be cached.
+                organization_id=project.organization_id,
+                version=release,
+            ).id,
+        )
+
+        if key in results:
+            first_seen, last_seen = results[key]
+            results[key] = (event.datetime, last_seen)
+        else:
+            results[key] = (event.datetime, event.datetime)
+
+    return results
+
+
+def repair_group_release_data(project, events):
+    attributes = collect_release_data(project, events).items()
+    for (group_id, environment, release_id), (first_seen, last_seen) in attributes:
+        instance, created = GroupRelease.objects.get_or_create(
+            project_id=project.id,
+            group_id=group_id,
+            environment=environment,
+            release_id=release_id,
+            defaults={
+                'first_seen': first_seen,
+                'last_seen': last_seen,
+            },
+        )
+
+        if not created:
+            instance.update(first_seen=first_seen)
+
+
+def collect_tsdb_data(project, events):
+    counters = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(int),
+        ),
+    )
+
+    sets = defaultdict(
+        lambda: defaultdict(
+            lambda: defaultdict(int),
```

Comment:
This is wrong, I'll fix when I add coverage below to the user tracking path. ("Static type systems don't catch bugs, they said…")